R/drive_publish.R

```diff
@@ -26,150 +32,119 @@ drive_publish <- function(file, ..., verbose = TRUE) {
-  ## we need to run again
-  if (isTRUE(x$published) && isTRUE(x$publishAuto)) {
-    response <- make_request(request, encode = "json")
-    proc_res <- process_response(response)
-  }
-
-  file_update$publish <- publish_tbl(proc_res)
-  invisible(file_update)
-}
-
-#' Check if Google Drive file is published
+#' @seealso Wraps the `revisions.update` endpoint:
+#'  * <https://developers.google.com/drive/v3/reference/revisions/update>
 #'
-#' @template file
+#' @template file-plural
+#' @param ... Name-value pairs to add to the API request body.
 #' @template verbose
 #'
 #' @template dribble-return
 #' @export
 #' @examples
 #' \dontrun{
-#' ## Upload file to check publication status
+#' #' ## Upload file to publish
 #' file <- drive_upload(R.home('doc/BioC_mirrors.csv'),
 #'   type = "spreadsheet")
 #'
-#' ## Check publication status
-#' drive_is_published(file)
-#'
 #' ## Publish file
-#' drive_publish(file)
-#'
-#' ## Check publication status agian
-#' drive_is_published(file)
+#' file <- drive_publish(file)
 #'
-#' ## Unpublish file
-#' drive_unpublish(file)
+#' ## Unplublish file
+#' file <- drive_unpublish(file)
 #'
 #' ## Clean up
 #' drive_rm(file)
 #' }
-drive_is_published <- function(file, verbose = TRUE) {
+drive_unpublish <- function(file, ..., verbose = TRUE) {
+  drive_change_publish(file = file, publish = FALSE, ..., verbose = verbose)
+}
+
+drive_change_publish <- function(file,
+                                 publish = TRUE,
+                                 ...,
+                                 verbose = TRUE) {
 
   file <- as_dribble(file)
   file <- confirm_some_files(file)
 
   mime_types <- purrr::map_chr(file$drive_resource, "mimeType")
-  if (!all(grepl("application/vnd.google-apps.", mime_types)) || is_folder(file)) {
+  if (!all(grepl("application/vnd.google-apps.", mime_types)) || any(is_folder(file))) {
     all_mime_types <- glue_data(file, "  * {name}: {mime_types}")
     stop_collapse(c(
-      "Only Google Drive type files can be published.",
-      "Your file(s) and type:",
+      "\nOnly Google Drive type files can be published.",
+      "Your file(s) are type:",
       all_mime_types,
       "Check out `drive_share()` to change sharing permissions."
     ))
   }
 
-  published <- purrr::map2(file$id, file$name, is_published_one, verbose = verbose)
-  file$publish <- published
-
+  params <- list(...)
+  params[["published"]] <- publish
+  params[["publishAuto"]] <- params[["publishAuto"]] %||% TRUE
+  params[["publishedOutsideDomain"]] <- params[["publishedOutsideDomain"]] %||% TRUE
+  params[["revisionId"]] <- "head"
+  params[["fields"]] <- "*"
+
+  revision_resource <- purrr::map(file$id,
+                                  change_publish_one,
+                                  params = params)
+  ## should we re-register the file here? technically when you publish,
+  ## the version increases by 1 (which is in drive_resource) so if
+  ## we don't re-register, the version will be incorrect.
```

Comment:
The `version` in the file resource is this:

> A monotonically increasing version number for the file. This reflects every change made to the file on the server, even those not visible to the user.

Based on an experiment I've done, this is NOT the same as the revision. And it doesn't increment by 1. I'm starting at a version of 7, even though the only thing I've done to the file is upload and publish. The revision is still at 1. But that's not exactly your point.

Perhaps we should call `drive_reveal(file, "publish")` and take care of it that way. Two good effects:

  * updates the metadata
  * reinforces modularity